[TSan] Clarify and enforce shadow end alignment

Camsyn opened the pull request:
In TSan, every `k` bytes of application memory (where `k = 8`) maps to a single shadow/meta cell. This design leads to two distinct outcomes when calculating the end of a shadow range using `MemToShadow(addr_end)`, depending on the alignment of `addr_end`:

  * **Exclusive End:** If `addr_end` is aligned (`addr_end % k == 0`), `MemToShadow(addr_end)` points to the first shadow cell *past* the intended range. This address is an exclusive boundary marker, not a cell to be operated on.
  * **Inclusive End:** If `addr_end` is not aligned (`addr_end % k != 0`), `MemToShadow(addr_end)` points to the last shadow cell that *is* part of the range (i.e., the same cell as `MemToShadow(addr_end - 1)`).

Different TSan functions have different expectations for whether the shadow end should be inclusive or exclusive. However, these expectations are not always explicitly enforced, which can lead to subtle bugs or reliance on unstated invariants.


The core of this patch is to ensure that functions ONLY requiring an **exclusive shadow end** behave correctly.

1.  Enforcing Existing Invariants:
    For functions like `MetaMap::MoveMemory` and `MapShadow`, the assumption is that the end address is always `k`-aligned. While this holds true in the current codebase (e.g., due to some external implicit conditions), this invariant is not guaranteed by the function's internal context. We add explicit assertions to make this requirement clear and to catch any future changes that might violate this assumption.

2.  Fixing Latent Bugs:
    In other cases, unaligned end addresses are possible, representing a latent bug. This was the case in `UnmapShadow`. The `size` of a memory region being unmapped is not always a multiple of `k`. When this happens, `UnmapShadow` would fail to clear the final (tail) portion of the shadow memory.

    This patch fixes `UnmapShadow` by rounding up the `size` to the next multiple of `k` before clearing the shadow memory. This is safe because the underlying OS `unmap` operation is page-granular, and the page size is guaranteed to be a multiple of `k`.

    Notably, this fix makes `UnmapShadow` consistent with its inverse operation, `MemoryRangeImitateWriteOrResetRange`, which already performs a similar size round-up.

In summary, this PR:

  - **Adds assertions** to `MetaMap::MoveMemory` and `MapShadow` to enforce their implicit requirement for k-aligned end addresses.
  - **Fixes a latent bug** in `UnmapShadow` by rounding up the size to ensure the entire shadow range is cleared. Two new test cases have been added to cover this scenario.
  - Removes a redundant assertion in `__tsan_java_move`.
  - Fixes an incorrect shadow end calculation introduced in commit 4052de6. The previous logic, while fixing an overestimation issue, did not properly account for `kShadowCell` alignment and could lead to underestimation.

Camsyn committed:
[TSan] Fix potentially problematic shadow end calculations

This is an improvement that enhances the robustness of the code.

Previously, the correct calculation of exclusive EndShadow relied on the
assumption that `addr_end % kShadowCell == 0`; however, in many current
usages, this assumption was not strictly guaranteed (although it did in
fact meet).

In addition, computing EndShadow does not require the corresponding
address to be AppMem; for example, HighAppEnd is not AppMem, but can
still be used to calculate EndShadow.

For example, for the AppMem range [0, 1), `s = MemToShadow(0)` is equal
to `MemToShadow(1)`. The previous logic would incorrectly deduce an
empty shadow range [s, s) while the correct shadow range should be
[s, s + kShadowSize * kShadowCnt) to cover all the related shadow memory
for the accessed cell.

This commit addresses this in two ways:
1. It introduces a dedicated utility function, i.e., `MemToEndShadow`,
    to correctly calculate the end of a shadow memory range, accounting
    for the memory cell granularity.
2. It replaces existing (and potentially incorrect) calculations of the
    shadow end with this new utility function.

Additionally, the previous commit 4052de6 resolved a problem with
overestimating the shadow end; it did not consider `kShadowCell` and
could therefore lead to underestimates. This is also corrected by
utilizing the `MemToEndShadow` function.

melver commented:
Is there a bug? If there's a bug, this needs a test. Otherwise, it's unclear what this is improving.

Camsyn commented:
> Is there a bug? If there's a bug, this needs a test. Otherwise, it's unclear what this is improving.

I acknowledge that "This is an improvement that enhances the robustness of the code", because the potential 'bugs' can hardly be triggered under the current implementation.

---
In tsan_interface_java.cpp:
```cpp
  DCHECK_EQ(dst % kHeapAlignment, 0);
  DCHECK_EQ(size % kHeapAlignment, 0);
  ...
  // Duplicated assertion
  DCHECK_NE(size, 0);
  ...
  RawShadow *dend = MemToShadow(dst + size);
```
It should be a bug if `kHeapAlignment != kShadowCell`, although they are both equal to 8 in the current implementation.

---
In tsan_rtl_access.cpp:
```cpp
  RawShadow* shadow_mem_end = reinterpret_cast<RawShadow*>(
      reinterpret_cast<uptr>(shadow_mem) + size * kShadowMultiplier - 1);
  if (!IsShadowMem(shadow_mem_end)) {
```
The calculation of `shadow_mem_end` did not consider `kShadowCell` and could therefore lead to underestimates.
E.g., for app mem [0, 1), it gives a shadow end `shadow_of(0) + 1`, while the real shadow end is `shadow_of(0) + 16`.

What's more, if `size == 0` (which is not allowed in its caller `MemoryAccessRange` but allowed in `MemoryAccessRangeT`), the current calculation of end is problematic.

---
In tsan_platform_linux.cpp:

TSan iterates the proc map and for each segment, 
`char *shadow_end = (char *)MemToShadow(segment.end);` is correct as the assumption of `segment.end % kShadowCell == 0` always holds. But do we need to guarantee this condition more explicitly?

---
In tsan_rtl.cpp:
```
void DontNeedShadowFor(uptr addr, uptr size) {
  ReleaseMemoryPagesToOS(reinterpret_cast<uptr>(MemToShadow(addr)),
                         reinterpret_cast<uptr>(MemToShadow(addr + size)));
}
```

`DontNeedShadowFor(addr, size)` is used in `ThreadFinish` to release the shadow related to the thread stack, and used in  `unmap(p, size)` to release the shadow allocated for `mmap`. 
What if `size % kShadowCell != 0`? I think it might lead to a missing cleanup of the tail shadow.

E.g., for `mmap(addr, size, ..)`,  TSan rounds up the `size` to poison the shadow, while in `munmap`, TSan does not (see `UnmapShadow`). If `MemToShadow(addr + size) % kPageSize == 0` and `MemToEndShadow(addr + size) > MemToShadow(addr + size)`, the page of the tail shadow cannot be released theoretically.

What's more, I found another potential bug in `DontNeedShadowFor` while analyzing: `DontNeedShadowFor` receives [beg, end] as parameters while TSan transfers [beg, end) as arguments, which might lead to page overreleasing.

In addition, another interface `MapShadow(addr, size)` shares the same problem.

Camsyn committed:
Remove XFAIL marker from 2 tests

Camsyn committed:
Use RoundUp on demand instead of a new util function

We don't always need to get the real shadow end, for example, some
shadow clear, if using the real shadow end, it will cause overcleaning
(i.e., clear [0, 1) makes [1, 8) inaccessible when kShadowCell == 8).

So when we do need to get the real end, use RoundUp in place.
For example, `unmap(addr, sz)` makes `[addr + sz, addr + PageSize)`
inaccessible, so we can safely clean up the full shadow (by getting
the real shadow end).

Camsyn committed:
add some assertions

Camsyn committed:
Remove all MemToEndShadow usages

thurstond commented on compiler-rt/test/tsan/java_heap_init2.cpp:
```diff
@@ -0,0 +1,32 @@
+// RUN: %clangxx_tsan -O1 %s -o %t && %run %t 2>&1 | FileCheck %s
```
Please add these test cases to a separate pull request. Since your fix is not committed yet, the tests would need an ["XFAIL" annotation](https://llvm.org/docs/TestingGuide.html#constraining-test-execution). Then, after the test cases are landed, this fix PR would update the test by removing the XFAIL. This two-step procedure makes it clear that 1) TSan was broken 2) this PR fixes it.

Camsyn replied:
If it was unfixed in your environment, that was an exception (and would you mind providing more information about the test failure, such as the triggering environment?); 

This relevant bug should indeed be fixed by this PR, and I have passed this test (Ubuntu24.04@x86_64) before I pushed the commit.

FYI, this test case manages to trigger a situation that `munmap` clears the meta incompletely due to the incorrect usage of an inclusive meta end. However, this bug should be fixed in this PR, as follows:
```cpp
  // This round-up gives the exclusive meta end.
  uptr size_for_meta = RoundUp(addr + size, kMetaShadowCell) - addr;
  ctx->metamap.ResetRange(thr->proc(), addr, size_for_meta, true);
```

Camsyn replied:
I submitted the testcase and fix together in several PRs before, so I did the same this time. 
But your suggestion is really a better practice, and I have put forward a separate PR #145472. :)
After that PR is merged, I will rebase this PR.

thurstond replied:
Thanks!

thurstond commented on compiler-rt/lib/tsan/rtl/tsan_rtl_access.cpp:
```diff
@@ -688,16 +688,17 @@ void MemoryAccessRangeT(ThreadState* thr, uptr pc, uptr addr, uptr size) {
-      reinterpret_cast<uptr>(shadow_mem) + size * kShadowMultiplier - 1);
-  if (!IsShadowMem(shadow_mem_end)) {
-    Printf("Bad shadow end addr: %p (%p)\n", shadow_mem_end,
+  uptr size1 =
```
Nit: please give a more descriptive name (e.g., rounded_size, aligned_size, etc.)

thurstond commented on compiler-rt/lib/tsan/rtl/tsan_rtl_access.cpp:
```diff
@@ -688,16 +688,17 @@ void MemoryAccessRangeT(ThreadState* thr, uptr pc, uptr addr, uptr size) {
+      (RoundUpTo(addr + size, kShadowCell) - RoundDownTo(addr, kShadowCell));
+  RawShadow* shadow_mem_end = shadow_mem + size1 / kShadowCell * kShadowCnt;
+  if (!IsShadowMem(shadow_mem_end - 1)) {
+    Printf("Bad shadow end addr: %p (%p)\n", shadow_mem_end - 1,
```
Nit: please add size1 to one of the Printfs

Camsyn committed:
Make var more descriptive

Camsyn committed:
Unify the var name to `rounded_size`

Camsyn commented:
> Consider reverting this commit, and then re-committing with the fix.
After debugging, I found that it is a behavior difference of `ReleaseMemoryPagesToOS` between IOS and Linux.

In IOS, it executes `madvise(beg, end, MADV_FREE)`, which tags the relevant pages as 'FREE' and does not release them immediately.
In Linux, it executes `madvise(beg, end, MADV_DONTNEED)`, releasing the pages immediately and setting them to 0.

Therefore, in IOS, ` __tsan_read1(&buf[0]);` still raises a race error even after `munmap` 'clears' the relevant shadow pages.

I think we need to change the test case to one of the following fixes:
1. Directly verify the page status rather than expect `__tsan_read1` not to raise an error (For an end-to-end testcase, this is too detailed).
2. Turn off the test case `munmap_clear_shadow.c`, because TSan does not provide a direct API query whether the shadow pages are cleared.

Do you think which fix is better, or can you give a better suggestion? @melver 

After confirming the fix, I will propose another Reland PR. Or for such a simple testcase-only fix, should I just fix-forward without revert?

Camsyn commented:
> 
> Turning off the test case is unacceptable - that'd be a regression.
> 
> Could you explain why this commit introduced the regression?

This triggered test case `munmap_clear_shadow.c` was supposed to be part of this PR, but was separated into my other PR #145472, adhering to the review suggestion.

I did pass the test case in my environment of Linux/x86_64, but lacked consideration of other platforms. 
Therefore, it caused the test case to still be triggered in IOS as the cleared shadow is not set to 0 after `ReleaseMemoryPagesToOS`.
> In IOS, it executes madvise(beg, end, MADV_FREE), which tags the relevant pages as 'FREE' and does not release them immediately.
In Linux, it executes madvise(beg, end, MADV_DONTNEED), releasing the pages immediately and setting them to 0.

melver commented:
> > Turning off the test case is unacceptable - that'd be a regression.
> > Could you explain why this commit introduced the regression?
> 
> This triggered test case `munmap_clear_shadow.c` was supposed to be part of this PR, but was separated into my other PR #145472, adhering to the review suggestion.
> 
> I did pass the test case in my environment of Linux/x86_64, but lacked consideration of other platforms. Therefore, it caused the test case to still be triggered in IOS as the cleared shadow is not set to 0 after `ReleaseMemoryPagesToOS`, and accordingly caused `__tsan_read1` to raise a race error in the test case, which is not my expectation.
> 
> > In IOS, it executes madvise(beg, end, MADV_FREE), which tags the relevant pages as 'FREE' and does not release them immediately.
> > In Linux, it executes madvise(beg, end, MADV_DONTNEED), releasing the pages immediately and setting them to 0.

Good point.

If this is a new test, feel free to disable it on iOS.
